File: sirius-core/src/main/java/org/starcoin/core/MockContext.java
package org.starcoin.core;

import java.util.HashMap;
import java.util.Map;

public class MockContext {

    public static final MockContext DEFAULT = new MockContext();

    private Map<String, Object> properites = new HashMap<>();

    public boolean containsKey(String key) {
        return properites.containsKey(key);
    }

    public Object get(String key) {
        return properites.get(key);
    }

    public boolean getAsBoolean(String key) {
        Object v = this.get(key);
        if (v == null) {
            return false;
        }
        return (Boolean) v;
    }

    public int getAsInt(String key) {
        Object v = this.get(key);
        if (v == null) {
            return 0;
        }
        return (Integer) v;
    }

    public MockContext put(String key, Object value) {
        properites.put(key, value);
        return this;
    }

    public <V> V getOrDefault(String key, V defaultValue) {
        return (V) properites.getOrDefault(key, defaultValue);
    }

    public <V> V getOrSet(String key, V setValue) {
        Object v = this.get(key);
        if (v == null) {
            this.properites.put(key, setValue);
            v = setValue;
        }
        return (V) v;
    }

    public static <T extends Mockable> T mock(Class<T> clazz, MockContext context) {
        try {
            T t = clazz.newInstance();
            t.mock(context);
            return t;
        } catch (InstantiationException | IllegalAccessException e) {
            throw new RuntimeException(e);
        }
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/MockContext.kt
package org.starcoin.sirius.core

import org.starcoin.sirius.crypto.CryptoKey
import org.starcoin.sirius.crypto.CryptoService
import java.util.*
import kotlin.reflect.KClass
import kotlin.reflect.full.companionObjectInstance

object MockContext {

    private val properites = HashMap<String, Any?>()
    var cryptoKey: CryptoKey by properites

    init {
        cryptoKey = CryptoService.dummyCryptoKey
    }

    fun containsKey(key: String): Boolean {
        return properites.containsKey(key)
    }

    operator fun get(key: String): Any? {
        return properites[key]
    }

    fun getAsBoolean(key: String): Boolean {
        val v = this[key] ?: return false
        return v as Boolean
    }

    fun getAsInt(key: String): Int {
        val v = this[key] ?: return 0
        return v as Int
    }

    fun put(key: String, value: Any): MockContext {
        properites[key] = value
        return this
    }

    fun <V> getOrDefault(key: String, defaultValue: V): V {
        return (properites as java.util.Map<String, Any>).getOrDefault(key, defaultValue) as V
    }

    fun <V> getOrSet(key: String, setValue: V): V {
        var v = this[key]
        if (v == null) {
            this.properites[key] = setValue
            v = setValue
        }
        return v as V
    }


    fun <T : SiriusObject> mock(clazz: KClass<T>): T {
        val companion = clazz.companionObjectInstance as SiriusObjectCompanion<T, *>
        return companion.mock()
    }
}
